### ISSUE TITLE
Not working out of box with kotlin plugin

### ISSUE THREAD

**Issue by sreich** [google/protobuf-gradle-plugin#100]
gradle version: 3.0
protobuf-gradle-plugin version: 0.8.0

having some problem with utilizing this in my project, it seems that although i can invoke the './gradlew generateProto' task myself, and then run a ':build', which would work just fine...it does not do this by itself.

i was a bit surprised, because the other examples i've seen, they don't need to specify a dependency on that task or anything like that. simply a dep on the plugin seems to suffice.

probably easier if i just give you a link to the project in question(gradle file's kinda big)

https://github.com/sreich/ore-infinium/blob/a88d5b3dee303973830b778a8b6430bdc85ef795/build.gradle#L147

ideally i would like a 'desktop:build' to execute that (or really, for any build task). specifying it in core should be enough, not sure why it isn't picking it up. what am i missing?

**Comment by sreich** [comment]
looks like i was able to fix this, although it's possibly a bit of a hack/workaround. maybe it doesn't like kotlin?

newer version that works https://github.com/sreich/ore-infinium/blob/627739e7274cebc01e76764837651b3c95473b9f/build.gradle

tl;dr:

added a 

compileKotlin {
dependsOn(':core:generateProto')
}
at project(":core")'s top level, that solved the issue of it not getting generated. the next issue to tackle was it not getting added to source sets

which i realized core/build.gradle overrides sourceSets.main.kotlin.srcDirs. so i added an entry of that to point to the build/generated/source/proto/.. dir

**Comment by zhangkun83** [comment]
The protobuf plugin makes `compileJava` depend on `generateProto`. It works for Java projects. IIUC in Kotlin projects, `compileJava` also depends on `compileKotlin`.

In your case `compileKotlin` uses the proto generated classes, but it doesn't have a dependency to `generateProto`, thus it failed.

Ideally, the protobuf plugin should do the same thing to `compileKotlin` as it does to `compileJava`, i.e., make it depend on the output of `generateProto`.

**Comment by nesterov-n** [comment]
@zhangkun83 Hello. I figured out that protobuf gradle plugin doesn't work out of box with kotlin android projects too.
I made a simple example https://github.com/nesterov-n/KotlinProtobufExample
`Message.java` is generated from `Message.proto` and can be used in java code. But using it in kotlin code finished with `Unresolved reference: Message`
In android project `generateProto`task is called before `compileKotlin` task, so workaround suggested by @sreich doesn't work in android kotlin-projects.
Can you help me with this issue? Maybe there is some android-specific workarounds to make kotlin-projects works with protobuf? Or maybe you have some thoughts how to improve protobuf gradle plugin? Ready to make PR with corrections

**Comment by Jacksgong** [comment]
@nesterov-n I have the same problem with you.

**Comment by Jacksgong** [comment]
@nesterov-n I have the same problem with you, then I have to execute `./gradlew extractDebugProto` before execute `./gradlew assembleDebug` to solve such problem.

**Comment by Jacksgong** [comment]
@nesterov-n You can get the temporary solution on [here](https://github.com/Jacksgong/grpc-android-kotlin) for Android with kotlin and protobuf.

**Comment by passsy** [comment]
I created a separate (java) module for the proto files and include it in my android project with Kotlin.

proto module `build.gradle`
```gradle
apply plugin: 'java'
apply plugin: 'com.google.protobuf'

dependencies {
    compile 'com.google.protobuf:protobuf-lite:3.0.0'
}

sourceCompatibility = "1.7"
targetCompatibility = "1.7"

sourceSets {
    main.java.srcDirs += 'src/main/javalite'
}

protobuf {
    protoc {
        artifact = 'com.google.protobuf:protoc:3.0.0'
    }
    plugins {
        javalite {
            artifact = 'com.google.protobuf:protoc-gen-javalite:3.0.0'
        }
    }
    generateProtoTasks.generatedFilesBaseDir = "${project.projectDir}/src/"
    generateProtoTasks {
        all().each { task ->
            task.builtins {
                remove java
            }
            task.plugins {
                javalite {}
            }
        }
    }
}


clean {
    delete "${protobuf.generatedFilesBaseDir}main/javalite"
}
```

app `build.gradle`
```gradle
dependencies {
    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    //...
    compile project(':proto')
}
```

`settings.gradle`
```gradle
include ':app', ':proto'
```

**Comment by while-loop** [comment]
I also had the same problem when using gRPC with Kotlin. Currently using Android Studio 3.0 beta7.
It worked fine as a Java app, but once converted to kotlin it wasn't able to find the resources in the build directory. 

I followed @passsy's suggestion and created a lib for the protos.


proto `build.gradle`
```gradle
apply plugin: 'java-library'
apply plugin: 'com.google.protobuf'

buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'com.google.protobuf:protobuf-gradle-plugin:0.8.3'
    }
}

def grpcVersion = '1.6.1'
def protobufVersion = '3.4.0'

dependencies {
    compile "com.google.protobuf:protobuf-java:${protobufVersion}"
    compile "io.grpc:grpc-okhttp:${grpcVersion}"
    compile "io.grpc:grpc-protobuf:${grpcVersion}"
    compile "io.grpc:grpc-stub:${grpcVersion}"
}

sourceSets {
    main {
        proto {
            srcDir '../remember-me/proto'
        }
    }
}

protobuf {
    protoc {
        artifact = "com.google.protobuf:protoc:${protobufVersion}"
    }
    plugins {
        grpc {
            artifact = "io.grpc:protoc-gen-grpc-java:${grpcVersion}"
        }
    }

    generateProtoTasks {
        ofSourceSet('main')*.plugins {
            grpc {}
        }
    }
}

clean {
    delete rootProject.buildDir
}
```

app `build.gradle`
```gradle
...
dependencies {
    compile "org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
    //...
    compile project(':proto')
}
```

`settings.gradle`
```gradle
include ':app', ':proto'
```

**Comment by macduy** [comment]
If you have grpc services, also add

```
dependencies {
    implementation 'io.grpc:grpc-protobuf-lite:1.5.0'
    implementation 'io.grpc:grpc-stub:1.5.0'
}
```

**Comment by macduy** [comment]
Based on @passsy and @while-loop's comments, this is the set up that works for me and generates grpc services too. `build.gradle`

```gradle
apply plugin: 'java'
apply plugin: 'com.google.protobuf'

dependencies {
    compile 'com.google.protobuf:protobuf-lite:3.0.0'
}

sourceSets {
    main {
        proto {
            srcDir '../PATH/TO/PROTOS'
        }
    }
}

protobuf {
    protoc {
        artifact = 'com.google.protobuf:protoc:3.3.0'
    }
    plugins {
        javalite {
            artifact = "com.google.protobuf:protoc-gen-javalite:3.0.0"
        }
        grpc {
            artifact = 'io.grpc:protoc-gen-grpc-java:1.5.0' // CURRENT_GRPC_VERSION
        }
    }
    generateProtoTasks {
        all().each { task ->
            task.builtins {
                // By default, java is always enabled. Remove it
                remove java
            }
            task.plugins {
                javalite {}
                grpc {
                    option 'lite'
                }
            }
        }
    }
}

dependencies {
    implementation 'io.grpc:grpc-protobuf-lite:1.5.0'
    implementation 'io.grpc:grpc-stub:1.5.0'
}

apply plugin: 'idea'

idea {
    module {
        sourceDirs += file("${projectDir}/build/generated/source/proto/main/javalite")
        sourceDirs += file("${projectDir}/build/generated/source/proto/main/grpc")
    }
}
```

**Comment by zpencer** [comment]
The following PRs add kotlin support:
https://github.com/google/protobuf-gradle-plugin/pull/196
https://github.com/google/protobuf-gradle-plugin/pull/201
